src/detect-tls-cert.c

```diff
@@ -0,0 +1,496 @@
+
+static int DetectTlsCertSetup(DetectEngineCtx *, Signature *, const char *);
+static void DetectTlsCertRegisterTests(void);
+static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms,
+        Flow *_f, const uint8_t _flow_flags,
+        void *txv, const int list_id);
+static int g_tls_cert_buffer_id = 0;
+
+/**
+ * \brief Registration function for keyword: tls_cert
+ */
+void DetectTlsCertRegister(void)
+{
+    sigmatch_table[DETECT_AL_TLS_CERT].name = "tls_cert";
+    sigmatch_table[DETECT_AL_TLS_CERT].desc = "content modifier to match the TLS certificate buffer";
+    sigmatch_table[DETECT_AL_TLS_CERT].url = DOC_URL DOC_VERSION "/rules/tls-keywords.html#tls-cert";
+    sigmatch_table[DETECT_AL_TLS_CERT].Match = NULL;
+    sigmatch_table[DETECT_AL_TLS_CERT].Setup = DetectTlsCertSetup;
+    sigmatch_table[DETECT_AL_TLS_CERT].Free = NULL;
+    sigmatch_table[DETECT_AL_TLS_CERT].RegisterTests = DetectTlsCertRegisterTests;
+
+    sigmatch_table[DETECT_AL_TLS_CERT].flags |= SIGMATCH_NOOPT;
+
+    DetectAppLayerInspectEngineRegister2("tls_cert", ALPROTO_TLS,
+            SIG_FLAG_TOCLIENT, TLS_STATE_CERT_READY,
+            DetectEngineInspectBufferGeneric, GetData);
+
+    DetectAppLayerMpmRegister2("tls_cert", SIG_FLAG_TOCLIENT, 2,
+            PrefilterGenericMpmRegister, GetData, ALPROTO_TLS,
+            TLS_STATE_CERT_READY);
+
+    DetectBufferTypeSetDescriptionByName("tls_cert", "TLS certificate");
+
+    g_tls_cert_buffer_id = DetectBufferTypeGetByName("tls_cert");
+}
+
+/**
+ * \brief This function setup the tls_cert modifier keyword
+ *
+ * \param de_ctx Pointer to the Detect Engine Context
+ * \param s      Pointer to the Signature to which the keyword belongs
+ * \param str    Should hold an empty string always
+ *
+ * \retval 0 On success
+ * \retval -1 on failure
+ */
+static int DetectTlsCertSetup(DetectEngineCtx *de_ctx, Signature *s,
+                              const char *str)
+{
+    DetectBufferSetActiveList(s, g_tls_cert_buffer_id);
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_TLS) != 0)
+        return -1;
+
+    return 0;
+}
+
+static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Flow *_f,
+        const uint8_t _flow_flags, void *txv, const int list_id)
+{
+    BUG_ON(det_ctx->inspect_buffers == NULL);
+    InspectionBuffer *buffer = &det_ctx->inspect_buffers[list_id];
+
+    if (buffer->inspect == NULL) {
+        SSLState *ssl_state = (SSLState *)_f->alstate;
+
+        if (TAILQ_EMPTY(&ssl_state->server_connp.certs)) {
+            return NULL;
+        }
+
+        SSLCertsChain *cert = TAILQ_FIRST(&ssl_state->server_connp.certs);
+        if (cert == NULL) {
+            return NULL;
+        }
+
+        InspectionBufferSetup(buffer, cert->cert_data, cert->cert_len);
+        InspectionBufferApplyTransforms(buffer, transforms);
+    }
+
+    return buffer;
+}
+
+#ifdef UNITTESTS
```

Comment:
can you move the tests into their own file in tests/ ?

Reply:
Sure :)